### PR TITLE
Accounts db replication design proposal

### PR DESCRIPTION
**Pull request by lijunwangs**
#### Problem
Validators fall behind the network when bogged down by heavy RPC load. This
seems to be due to a combination of CPU load and lock contention caused by
serving RPC requests. The most expensive RPC requests involve account scans.

#### Summary of Changes

The AccountsDb replication design proposal is described.

Fixes #

### CODE REVIEW

File: docs/src/proposals/accounts-db-replication.md [hunk 1]
```diff
@@ -0,0 +1,189 @@
...
+levels.
+
+### Solana RPC Node
+A new node named solana-rpc-node will be introduced whose main responsibility is to maintain
+the AccountsDb replica. The RPC node or replica node is used interchangebly in this document.
+It will be a separate exectuable from the validator.
+
+The replica consists of the following major components.
+
+The `ReplRpcUpdatedSlotsRequestor`, this service is responsible for peridically sending the
```
Comment (CriesofCarrots) [inline docs/src/proposals/accounts-db-replication.md:56]:
```suggestion
The `ReplRpcUpdatedSlotsRequestor`: this service is responsible for periodically sending the
```

File: docs/src/proposals/accounts-db-replication.md [hunk 2]
```diff
@@ -0,0 +1,189 @@
...
+### Consistency Model
+The AccountsDb information is replicated asynchronously from the main validator to the replica.
+When a query against the replica's AccountsDb is made, the replica may not have the latest
+information of the latest slot. In this regard, it will be eventually consistent. However, for
+a particular slot, the information provided is consistent with the that of its peer validator
+for commitment levels confirmed and finalized. For V1, we only support queries at these two
+levels.
+
+### Solana RPC Node
+A new node named solana-rpc-node will be introduced whose main responsibility is to maintain
```
Comment (CriesofCarrots) [inline docs/src/proposals/accounts-db-replication.md:50]: I tend to think `solana-rpc-node` is a little too generic. Perhaps `solana-accounts-replica-node`?

File: docs/src/proposals/accounts-db-replication.md (hunk 1 above)
Comment (CriesofCarrots) [reply docs/src/proposals/accounts-db-replication.md:56]: Also, let's spell out "replica". Do we need Rpc?
```suggestion
The `ReplicaUpdatedSlotsRequestor`, this service is responsible for peridically sending the
```

Same suggestion for all following service, request, and struct names below

File: docs/src/proposals/accounts-db-replication.md [hunk 3]
```diff
@@ -0,0 +1,189 @@
...
+request `ReplRpcUpdatedSlotsRequest` to its peer validator or replica for the latest slots.
+It specifies the latest slot (last_replicated_slot) for which the replica has already
+fetched the accounts information for. This maintains the ReplWorkingSlotSet and manages
+the lifecycle of BankForks, BlockCommitmentCache (for the highest confirmed slot) and
+the optimistically confirmed bank.
+
+The `ReplRpcUpdatedSlotsServer`, this service is responsible for serving the
+`ReplRpcUpdatedSlotsRequest` and sends the `ReplRpcUpdatedSlotsResponse` back to the requestor.
+The response consists of a vector of new slots the validator knows of which is later than the
+specified last_replicated_slot. This services also runs in the main validator. This service
```
Comment (CriesofCarrots) [inline docs/src/proposals/accounts-db-replication.md:66]:
```suggestion
specified last_replicated_slot. This service also runs in the main validator. This service
```

File: docs/src/proposals/accounts-db-replication.md [hunk 4]
```diff
@@ -0,0 +1,189 @@
...
+a particular slot, the information provided is consistent with the that of its peer validator
+for commitment levels confirmed and finalized. For V1, we only support queries at these two
+levels.
+
+### Solana RPC Node
+A new node named solana-rpc-node will be introduced whose main responsibility is to maintain
+the AccountsDb replica. The RPC node or replica node is used interchangebly in this document.
+It will be a separate exectuable from the validator.
+
+The replica consists of the following major components.
```
Comment (CriesofCarrots) [inline docs/src/proposals/accounts-db-replication.md:54]:
```suggestion
The replica consists of the following major components:
```

File: docs/src/proposals/accounts-db-replication.md [hunk 5]
```diff
@@ -0,0 +1,189 @@
...
+the existing `AccountMeta` in addition to the account data length in bytes.
+
+The `ReplRpcAccountsServer`, this service is reponsible for serving the `ReplRpcAccountsRequest`
+and sends `ReplRpcAccountsResponse` to the requestor. The response contains the count of the
+ReplAccountInfo and the vector of ReplAccountInfo. This service runs both in the validator
+and the replica relaying replication information. The server can stream the account information
+from its AccountCache or from the storage if already flushed. This is similar to how a snapshot
+package is created from the AccountsDb with the difference that the storage does not need to be
+flushed to the disk before streaming to the client. If the account data is in the cache, it can
+be directly streamed. Care must be taken to avoid the account data for a slot get cleaned while
```
Comment (CriesofCarrots) [inline docs/src/proposals/accounts-db-replication.md:82]:
```suggestion
be directly streamed. Care must be taken to avoid account data for a slot being cleaned while
```

File: docs/src/proposals/accounts-db-replication.md [hunk 6]
```diff
@@ -0,0 +1,189 @@
...
+serving the streaming. When attempting a replication of a slot, if the slot is already cleaned
+up with accounts data cleaned as result of update in later rooted slots, the replica should 
+forsake this slot and try the later uncleaned root slot.
+
+During replication we also need to replicate the information of accounts
+who have been cleaned up due to zero lamports, i.e. we need to be able to tell the difference
+for the account in a given slot: it is not updated and hence no storage entry in that slot and
+the accounts have 0 lamports and have been cleaned up through the history. We may record this
```
Comment (CriesofCarrots) [inline docs/src/proposals/accounts-db-replication.md:90]:
```suggestion
During replication we also need to replicate the information of accounts
that have been cleaned up due to zero lamports, i.e. we need to be able to tell the difference
between an account in a given slot which was not updated and hence has no storage entry in that slot, and
tone that holds 0 lamports and has been cleaned up through the history. We may record this
```

File: docs/src/proposals/accounts-db-replication.md [hunk 7]
```diff
@@ -0,0 +1,189 @@
...
+via some "Tombstone" mechanism -- recording the dead accounts cleaned up for a slot. The
+tombstones themselves can be removed after exceeding retention period expressed as epochs.
+And attempt to replicate slots with tombstones removed will fail and the replica should skip
```
Comment (CriesofCarrots) [inline docs/src/proposals/accounts-db-replication.md:93]:
```suggestion
Any attempt to replicate slots with tombstones removed will fail and the replica should skip
```

File: docs/src/proposals/accounts-db-replication.md [hunk 8]
```diff
@@ -0,0 +1,189 @@
...
+this slot and try later ones.
+
+The `JsonRpcAccountsService`, this is the RPC service serving client requests for account
+information. The existing JsonRpcService serves other client calls than AccountsDb ones.
+The replica node only serves the AccountsDb calls.
+
+The existing JsonRpcService requires `BankForks`, `OptimisticallyConfirmedBank` and
+`BlockCommitmentCache` to load the Bank. The JsonRpcAccountsService will need to use 
+information obtained from ReplRpcUpdatedSlotsResponse to construct the AccountsDb. Question,
+do we still need the Bank in the replica?
```
Comment (CriesofCarrots) [inline docs/src/proposals/accounts-db-replication.md:103]: Is this question still relevant?

File: docs/src/proposals/accounts-db-replication.md [hunk 9]
```diff
@@ -0,0 +1,189 @@
...
+### Compatibility Consideration
+
+For protocol compatiblilty considerations, all the requests have the replication version which is
+initially set to 1. Alternatively, we can use the validator's version. The RPC server side
+shall check the request version and fail if it is not supported.
+
+### Replication Setup
+To limit adverse effects on the validator and the replica due to replication, they can be configured
+with a list of replica nodes which can form a replication pair with it. And the replica node is
+configured with the validator which can serve its requests.
```
Comment (CriesofCarrots) [inline docs/src/proposals/accounts-db-replication.md:119]: This will just be one replica, right?

File: docs/src/proposals/accounts-db-replication.md [hunk 10]
```diff
@@ -0,0 +1,189 @@
...
+- getTokenAccountBalance
+- getTokenSupply
+- getTokenLargestAccounts
+- getTokenAccountsByOwner
+- getTokenAccountsByDelegate
+
+Following APIs are not included:
+
+- getInflationReward
+- getEpochSchedule
```
Comment (CriesofCarrots) [inline docs/src/proposals/accounts-db-replication.md:155]: This is in the `included` list. Remove?

File: docs/src/proposals/accounts-db-replication.md [hunk 11]
```diff
@@ -0,0 +1,189 @@
...
+with a list of replica nodes which can form a replication pair with it. And the replica node is
+configured with the validator which can serve its requests.
+
+
+### Fault Tolerance
+The main responsibility of making sure the replication is tolerant of faults lies with the replica.
+In case of request failures, the replica shall retry the requests. 
+
+
+### Interface
```
Comment (CriesofCarrots) [inline docs/src/proposals/accounts-db-replication.md:127]: I've started to separate these methods in rpc, and I think we should actually have 3 lists:
- JsonRpcAccountsService-only APIs
- JsonRpcService-only APIs
- APIs supported by both

Currently, most of the APIs that only require a Bank are in the `included` list, but a few are not: getSlotLeader, getSlotLeaders, and getBlockProduction
There's no reason that all of these couldn't be supported by both replicas and full API nodes; none of them are particularly heavy. And there is likely existing `Processed` commitment use with some of them, like `getFees` and `getRecentBlockhash`

Comment (CriesofCarrots) [reply docs/src/proposals/accounts-db-replication.md:127]: I'll push a PR shortly with rpc re-org along these lines and you can see what you think, @lijunwangs

File: docs/src/proposals/accounts-db-replication.md (hunk 2 above)
Comment (lijunwangs) [reply docs/src/proposals/accounts-db-replication.md:50]: Agreed on solana-rpc-node is a little too generic. solana-accounts-replica-node is a little too wordy with 4 words. How about solana-replica-node this keep it opens for other replications in the future.

File: docs/src/proposals/accounts-db-replication.md (hunk 9 above)
Comment (lijunwangs) [reply docs/src/proposals/accounts-db-replication.md:119]: Yes

File: docs/src/proposals/accounts-db-replication.md (hunk 10 above)
Comment (lijunwangs) [reply docs/src/proposals/accounts-db-replication.md:155]: Will do

File: docs/src/proposals/accounts-db-replication.md (hunk 8 above)
Comment (lijunwangs) [reply docs/src/proposals/accounts-db-replication.md:99]: Removed

File: docs/src/proposals/accounts-db-replication.md (hunk 7 above)
Comment (lijunwangs) [reply docs/src/proposals/accounts-db-replication.md]: Fixed

File: docs/src/proposals/accounts-db-replication.md (hunk 6 above)
Comment (lijunwangs) [reply docs/src/proposals/accounts-db-replication.md]: Done. Thanks

File: docs/src/proposals/accounts-db-replication.md (hunk 5 above)
Comment (lijunwangs) [reply docs/src/proposals/accounts-db-replication.md]: Done Thanks

File: docs/src/proposals/accounts-db-replication.md (hunk 4 above)
Comment (lijunwangs) [reply docs/src/proposals/accounts-db-replication.md]: Done

File: docs/src/proposals/accounts-db-replication.md (hunk 3 above)
Comment (lijunwangs) [reply docs/src/proposals/accounts-db-replication.md]: Done

File: docs/src/proposals/accounts-db-replication.md (hunk 1 above)
Comment (lijunwangs) [reply docs/src/proposals/accounts-db-replication.md]: Done

File: docs/src/proposals/accounts-db-replication.md [hunk 12]
```diff
@@ -0,0 +1,183 @@
...
+
+### Consistency Model
+The AccountsDb information is replicated asynchronously from the main validator to the replica.
+When a query against the replica's AccountsDb is made, the replica may not have the latest
+information of the latest slot. In this regard, it will eventually be consistent. However, for
+a particular slot, the information provided is consistent with that of its peer validator
+for commitment levels confirmed and finalized. For V1, we only support queries at these two
+levels.
+
+### Solana RPC Node
```
Comment (CriesofCarrots) [inline docs/src/proposals/accounts-db-replication.md:49]:
```suggestion
### Solana Replica Node
```

File: docs/src/proposals/accounts-db-replication.md [hunk 13]
```diff
@@ -0,0 +1,183 @@
...
+The `ReplicaUpdatedSlotsServer`: this service is responsible for serving the
+`ReplicaUpdatedSlotsRequest` and sends the `ReplicaUpdatedSlotsResponse` back to the requestor.
+The response consists of a vector of new slots the validator knows of which is later than the
+specified last_replicated_slot. This service also runs in the main validator. This service
+gets the slots for replication from the BankForks, BlockCommitmentCache and OptimiscallyConfirmBank.
+
+The `ReplicaAccountsRequestor`: this service is responsible for sending the request
+`ReplicaAccountsRequest` to its peer validator or replica for the `ReplAccountInfo` for a
+slot for which it has not completed accounts db replication. The `ReplAccountInfo` contains
+the `ReplAccountMeta`, Hash and the AccountData. The `ReplAccountMeta` contains info about
```
Comment (CriesofCarrots) [inline docs/src/proposals/accounts-db-replication.md:72]:
```suggestion
`ReplicaAccountsRequest` to its peer validator or replica for the `ReplicaAccountInfo` for a
slot for which it has not completed accounts db replication. The `ReplicaAccountInfo` contains
the `ReplicaAccountMeta`, Hash and the AccountData. The `ReplicaAccountMeta` contains info about
```

File: docs/src/proposals/accounts-db-replication.md (hunk 9 above)
Comment (CriesofCarrots) [reply docs/src/proposals/accounts-db-replication.md:115]: Let's update this section to reflect that, then.

File: docs/src/proposals/accounts-db-replication.md [hunk 14]
```diff
@@ -0,0 +1,183 @@
...
+flushed to the disk before streaming to the client. If the account data is in the cache, it can
+be directly streamed. Care must be taken to avoid account data for a slot being cleaned while
+serving the streaming. When attempting a replication of a slot, if the slot is already cleaned
+up with accounts data cleaned as result of update in later rooted slots, the replica should
+forsake this slot and try the later uncleaned root slot.
+
+During replication we also need to replicate the information of accounts
+that have been cleaned up due to zero lamports, i.e. we need to be able to tell the difference
+between an account in a given slot which was not updated and hence has no storage entry in that slot, and one that holds 0 lamports and has been cleaned up through the history. We may record this via
+some "Tombstone" mechanism -- recording the dead accounts cleaned up fora slot. The tombstones themselves can be removed after exceeding the retention period expressed as epochs. Any attempt to replicate slots with tombstones removed will fail and the replica should skip this slot and try later ones.
```
Comment (CriesofCarrots) [inline docs/src/proposals/accounts-db-replication.md:90]: Wrap lines?

File: docs/src/proposals/accounts-db-replication.md (hunk 8 above)
Comment (CriesofCarrots) [reply docs/src/proposals/accounts-db-replication.md:99]: Missing a commit? Doesn't seem removed.

**Review by CriesofCarrots — APPROVED** [review]
Lgtm, with nits

File: docs/src/proposals/accounts-db-replication.md (hunk 14 above)
Comment (lijunwangs) [reply docs/src/proposals/accounts-db-replication.md]: done
